### PR TITLE
implement autograd function trace

### PR DESCRIPTION
**Pull request by chenyuntc**
implement autograd  function   `trace` in linalg, and add it to Variable.
also add tests

### CODE REVIEW

File: torch/autograd/_functions/linalg.py [hunk 1]
```diff
@@ -41,4 +41,12 @@ def forward(self, input):
...
-# TODO: trace
+
+class Trace(Function):
+
+    def forward(self, input):
+        self.size = input.size()
+        return input.new((input.trace(),))
+
+    def backward(self, grad_output):
+        return grad_output[0] * (torch.eye(*self.size)).type_as(grad_output)
```
Comment (apaszke) [inline torch/autograd/_functions/linalg.py]: I think it's better to expand the `grad_output` and call `torch.diag` on it. This implementation will do a large CPU -> GPU transfer, just to create the identity matrix.

**Review by apaszke — CHANGES_REQUESTED** [review]

**Comment by Haoyanlong** [comment]
it's too late here, I'll fix it tomorrow.

**Comment by apaszke** [comment]
Can you please fix the lint error? Looks good now.

**Review by apaszke — APPROVED** [review]

**Comment by soumith** [comment]
i've fixed the lint

**Review by soumith — APPROVED** [review]

File: torch/autograd/_functions/linalg.py [hunk 2]
```diff
@@ -41,4 +41,12 @@ def forward(self, input):
...
-# TODO: trace
+
+class Trace(Function):
+
+    def forward(self, input):
+        self.size = input.size()
+        return input.new((input.trace(),))
+
+    def backward(self, grad_output):
+        return torch.diag(grad_output.expand(self.size[0]))
```
Comment (fmassa) [inline torch/autograd/_functions/linalg.py]: `trace` supports non-square matrices. Maybe we should support that as well?

**Comment by soumith** [comment]
Over here, I wrote a trace Function that correctly computes it for non-square matrices as well:
https://discuss.pytorch.org/t/function-for-trace/1015/2?u=smth

```
class CustomTrace(torch.autograd.Function):
    
    def forward(self, input):
        self.isize = input.size()
        return input.new([torch.trace(input)])
    
    def backward(self, grad_output):
        isize = self.isize
        grad_input = grad_output.new(isize).copy_(torch.eye(*isize))
        grad_input.mul_(grad_output[0])
        return grad_input
```

**Comment by apaszke** [comment]
I really think we should somehow avoid the CPU -> GPU copy... I think that using some smart stride tricks we could zero the gradient and only fill the diagonal (tensor that has stride[0] = size[1] + 1, and appropriate size will be a 1D view on the diagonal)

**Comment by fmassa** [comment]
The only annoying thing is that we calling `torch.eye` generates it on the CPU, and I'm not sure we already have a CUDA support for `torch.eye`.
We can get the `eye` function by something like
```python
a = torch.zeros(3, 2)
b = a.view(-1)
b[::3] = 1
print(a)
```

**Comment by apaszke** [comment]
Yeah, that's exactly what I wanted to do. Create a zeroed grad output of appropriate size, take a square slice, and fill the view of the diagonal with grad output

**Comment by chenyuntc** [comment]
```python
class Trace(Function):

    def forward(self, input):
        self.isize = input.size()
        return input.new((input.trace(),))

    def backward(self, grad_output):
        isize = self.isize
        grad_input = grad_output.new(isize).zero_()
        grad_input.view(-1)[::(isize[1] + 1)] = grad_output[0]
        return grad_input
```
is this implementation ok?

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks!
